presto-main/src/main/java/io/prestosql/operator/scalar/time/TimeToTimestampCast.java

```diff
@@ -57,7 +58,13 @@ public static LongTimestamp castToLong(
             @SqlType("time(sourcePrecision)") long time)
     {
         long epochMicros = cast(sourcePrecision, targetPrecision, session, time);
-        return new LongTimestamp(epochMicros, (int) (time % PICOSECONDS_PER_MICROSECOND));
+        int picosOfMicro = toIntExact(time % PICOSECONDS_PER_MICROSECOND);
+        picosOfMicro = toIntExact(round(picosOfMicro, toIntExact(12 - targetPrecision)));
+        if (picosOfMicro == PICOSECONDS_PER_MICROSECOND) {
+            epochMicros++;
+            picosOfMicro = 0;
+        }
+        return new LongTimestamp(epochMicros, picosOfMicro);
```

Comment:
This is kind of a roundabout way to do it. There's no point in rounding picosOfMicro twice (once in `cast()`, again in this new code).

Here's an alternate implementation that also fixes the testing issue: https://github.com/prestosql/presto/pull/5742